## The Promise.all() Method

The Promise.all() method accepts a single argument, which is an iterable (such as an array) of promises to monitor, and returns a promise that is resolved only when every promise in the iterable is resolved. The returned promise is fulfilled when every promise in the iterable is fulfilled, as in this example:

```js
let promise1 = Promise.resolve(42);

let promise2 = new Promise((resolve,reject)=>{ 
	resolve(43);
});

let promise3 = new Promise((resolve,reject)=>{ 
	setTimeout(() => {
        resolve(44);
    }, 100);
});

let promise4 = Promise.all([promise1,promise2,promise3]);

promise4.then(value=>{ 
	console.log(Array.isArray(value)); // true
	console.log(value[0]); // 42
	console.log(value[1]); // 43
	console.log(value[2]); // 44
});
```

### When to Use Promise.all()
You’ll want to use Promise.all() in any situation where you are waiting for multiple promises to fulfill, and any one failure should cause the entire operation to fail. Here are some common use cases for Promise.all().

- Processing Multiple Files Together
```js
// uses the Node.js promises-based filesystem API to read multiple files in parallel
import { readFile } from "node:fs/promises";

function readFiles(filenames) {
	return Promise.all(filenames.map(filename => readFile(filename, "utf8")))
}

readFiles([
    "file1.json",
    "file2.json"
]).then(fileContents=>{
// parse JSON data
const data = fileContents.map(fileContent => JSON.parse(fileContent));

// process as necessary
console.log(data);
}).catch((reason) => { 
	console.error(reason.message);
});
```

- Calling Multiple Dependent Web Service APIs
```js
const API_BASE = "https://jsonplaceholder.typicode.com"; 
function fetchUserData(userId) {
	const urls = [ 
		`${API_BASE}/users/${userId}/posts`,
		`${API_BASE}/users/${userId}/albums`
	];
	
	return Promise.all(urls.map(url => fetch(url)));
}

fetchUserData(1).then(responses=>{ 
	return Promise.all(
        responses.map(response => {
			if (response.ok) {
				return response.json();
			} else {  
				return Promise.reject(
					new Error(`Unexpected status code: ${response.status} ${response.statusText} for ${response.url}`)); 
			}
		 })
	 )
 }).then(([posts,albums])=>{
    // process your data as necessary
    console.log(posts);
    console.log(albums);
}).catch(reason => console.error(reason.message));
```

- Creating Artificial Delays
```js
// delay something from happening too quickly

function delay(milliseconds) {
	return new Promise((resolve) => {
		setTimeout(() => {
			resolve();	
		}, milliseconds)	
	})
}

function fetchUserData(userId) {
	const urls = [ 
		`${API_BASE}/users/${userId}/posts`,
		`${API_BASE}/users/${userId}/albums`
	];
	
	return Promise.all([
		...urls.map(url => fetch(url)),
		delay(2500)
	]).then((results) => {
		 // strip off the undefined result from delay()
		return results.slice(0, results.length - 1);
	});
}
```

If you want your operation to continue even if one promise is rejected? That’s where `Promise.allSettled()` is the better choice.

## The Promise.allSettled() Method

The Promise.allSettled() method is a slight variation of Promise.all() where the method waits until all promises in the specified iterable are settled, regardless of whether they are fulfilled or rejected. The return value of Promise.allSettled() is always a promise that is fulfilled with an array of result objects.

The result object for a fulfilled promise has two properties: 
	• status - always set to the string fulfilled
	• value - the fulfillment value of the promise  
 
For a rejected promise, there are also two properties on the result object:
	• status - always set to the string rejected 
	• reason - the rejection value of the promise

```js
let promise1=Promise.resolve(42);

let promise2=Promise.reject(43);

let promise3=new Promise((resolve,reject)=>{ 
	setTimeout(() => {
        resolve(44);
    }, 100);
}); 

let promise4=Promise.allSettled([promise1,promise2,promise3]);

promise4.then((results) => {
	console.log(Array.isArray(results)); // true
	console.log(results[0]); // {status: fulfilled, value: 42}
	console.log(results[1]); // {status: rejected, value: 43}
	console.log(results[2]); // {status: fulfilled, value: 44}
})
```

The Promise.allSettled() method can be used in a lot of the same situations as Promise.all(); however, it is best suited for when you want to ignore rejections, handle rejections differently, or allow partial success. Here are some common use cases for Promise.allSettled().

- Processing Multiple Files Separately
	There are also some cases where working on multiple files separately means you don’t need to stop the entire operation if one fails; you go ahead and complete the successful operations and then log the failed ones to retry later.

```js
import { readFile,writeFile } from"node:fs/promises";

// or any operation on the files
function transformText(text) {
	return text.split('').reverse().join('');
}

function transformFiles(filenames) {
	return Promise.allSettled(
		filenames.map(filename => 
			readFile(filename, 'utf-8')
			.then(transformText)
			.then(newText => writeFile(filename, newText))
			.catch(reason => {
				reason.filename = filename;
				return Promise.reject(reason);
			})
		)
	)
}

transformFiles(['file1.txt', 'file2.txt'])
.then(results => {
	// get failed results
	const failedResults = results.filter(result => result.status === 'rejected');

	if (failedResults.length) {
		console.error('The following files were not transformed successfully:');
		console.error('');
	
		failedResults.forEach(failedResult => {
			console.error(failedResult.reason.filename);
			console.error(failedResult.reason.message);
			console.error('');
		})
	} else {
		console.log('All files transformed successfully.');
	}
})

```

- Calling Multiple Independent Web Service APIs


## The Promise.any() Method
The Promise.any() method also accepts an iterable of promises and returns a fulfilled promise when any of the passed-in promises are fulfilled. The operation short-circuits as soon as one of the promises is fulfilled. (This is the opposite of Promise.all(), where the operation short-circuits as soon as one promise is rejected.)

```js
let promise1=Promise.reject(43);

let promise2=Promise.resolve(42);

let promise3=new Promise((resolve,reject)=>{ 
	setTimeout(() => {
        resolve(44);
    }, 100);

});  
let promise4=Promise.any([promise1,promise2,promise3]); 

promise4.then(value=>console.log(value)); // 42
```

Even though the first promise (promise1) in this example is rejected, the call to Promise.any() succeeds because the second promise (promise2) is fulfilled. The result of the third promise (promise3) is discarded. If all of the promises passed to Promise.any() are rejected, then the returned promise is rejected with an AggregateError. An AggregateError is an error that represents multiple errors stored in an errors property.

```js
let promise1=Promise.reject(43);

let promise2=Promise.reject(42);

let promise3=new Promise((resolve,reject)=>{ 
	setTimeout(() => {
        reject(44);
    }, 100);

});  
let promise4=Promise.any([promise1,promise2,promise3]); 

promise4.catch(reason => {
 // Runtime dependent error message
 console.log(reason.message);
 // output rejection values
 console.log(reason.errors) // [43, 42, 44]
});
```

The Promise.any() method is best used in situations where you want any one of the promises to fulfill and you don’t care how many others reject unless they all reject. Here are some situations where you might want to use Promise.any().

- Executing Hedged Requests
	As defined in The Tail at Scale2, a hedged request is one where the client makes requests to multiple servers and accepts the response from the first that replies. This is helpful in situations where the client needs the lowest latency possible, and there are server resources devoted to managing the extra load and deduplicating responses

```js
const HOSTS=["api1.example.com", "api2.example.com"];

function hedgedFetch(endpoint){ 
	return Promise.any(
		HOSTS.map(hostname => fetch(`https://${hostname}${endpoint}`))
	);
}

hedgedFetch("/transactions")  
.then(transactions => console.log(transations))
.catch(reason => console.error(reason.message));
```

- Using the Fastest Response in a Service Worker
	Web pages that use service workers often have their choice of where to load data from: the network or from the cache. In some cases, a network request might actually be faster than loading from cache, and so you may want to use Promise.any() to choose the faster of the responses

## The Promise.race() Method

The Promise.race() method provides a slightly different take on monitoring multiple promises. This method also accepts an iterable of promises to monitor and returns a promise, but the returned promise is settled as soon as the first promise is settled. Instead of waiting for all promises to be resolved like the Promise.all() method or short-circuiting only for the first resolved promise like Promise.any(), the Promise.race() method returns an appropriate promise as soon as any promise in the array is settled

```js
let promise1=Promise.resolve(42);

let promise2= new Promise((resolve) => {
	resolve(43);
});

let promise3=new Promise((resolve,reject)=>{ 
	setTimeout(() => {
        reject(44);
    }, 100);
});  
let promise4=Promise.race([promise1,promise2,promise3]); 

promise4.then(value => console.log(value)); // 42

```

The Promise.race() method is best used in situations where you want to be able to short-circuit the completion of a number of different promises

- Establishing a Timeout for an Operation
	While the fetch() function has a lot of helpful functionality, one thing it doesn’t do is manage a timeout for a given request; a request will happily hang until the request completes one way or another. You can easily create a wrapper method to add a timeout to any request by using Promise.race()

```js
function timeout(milliseconds) {
	return new Promise((resolve, reject) => {
		setTimeout(() => {
			reject(new Error('Request timed out.'));
		}, milliseconds)
	})
}

function fetchWithTimeout(...args){ 
	return Promise.race([
        fetch(...args),
        timeout(5000)
    ]);
}

const API_URL="https://jsonplaceholder.typicode.com/users";

fetchWithTimeout(API_URL)  
.then(response => response.json())  
.then(users => console.log(users))
.catch(reason => console.error(reason.message));
```

Note: Keep in mind that even though fetchWithTimeout() will reject if a request takes longer than the specified timeout, the request will not be cancelled. It will continue waiting for a response behind-the-scenes even though the response will be ignored.

### Summary

For times when you want to monitor and respond to multiple promises at the same time, JavaScript provides several methods. Each method behaves slightly differently, but all allow you to run promises in parallel and respond to them as a group:

- Promise.all() - the returned promise is fulfilled when all of the promises are fulfilled, and the returned promise is rejected when any promise is rejected.

- Promise.allSettled() - the returned promise is always fulfilled with an array of results from the promise, and the returned promise is never rejected.

- Promise.any() - the returned promise is fulfilled when the first promise is fulfilled, and the returned promise is rejected when all of the promises are rejected.

- Promise.race() - the returned promise is fulfilled when the first promise to settle is fulfilled, and the returned promise is rejected when the first promise to settle is rejected.

Each of these methods is appropriate for different use cases, and it’s up to you to decide which is appropriate in any situation.